gpu_memory_management/test/test.py

```diff
@@ -7,68 +7,90 @@
-    a = cp.ones(5)
+    cp.cuda.Device(0).use()
+    a = cp.array([3.14, 3.14, 3.14])
     res = gpuMemManagement.test_if_real_cupy_reinterpret_ptr_is_a_gpu_array(a.data.ptr)
 
     assert(res == 1)
 
 #test 5 : test if the pointer of a custom cupy pointer attributes is a device pointer
 def test_if_custom_cupy_reinterpret_ptr_is_a_gpu_array():
-    a = cp.ones(5)
+    cp.cuda.Device(0).use()
+    a = cp.array([3.14, 3.14, 3.14])
     b = cupy_ref.Custom_Cupy_Ref(ptr = a.data.ptr, size = a.size)
     res = gpuMemManagement.test_if_custom_cupy_reinterpret_ptr_is_a_gpu_array(b)
 
     assert(res == 1)
 
-#test 6 : copy array of ones from real cupy to cpu
-def test_copy_cupy_of_ones_to_cpu():
-    a = cp.ones(5)
-    b = gpuMemManagement.test_copy_cupy_of_ones_to_cpu(a.data.ptr, a.size)
+#test 6 : copy array of float from real cupy to cpu
+def test_copy_real_cupy_to_cpu():
+    cp.cuda.Device(0).use()
+    a = cp.array([3.14, 3.14, 3.14])
+    b = gpuMemManagement.test_copy_real_cupy_to_cpu(a.data.ptr, a.size)
 
     assert(b == 1)
 
-#test 7 : copy array of ones from custom cupy to cpu
-def test_copy_custom_cupy_of_ones_to_cpu():
-    a = cp.ones(5)
+#test 7 : copy array of float from custom cupy to cpu
+def test_copy_custom_cupy_to_cpu():
+    cp.cuda.Device(0).use()
+    a = cp.array([3.14, 3.14, 3.14])
     b = cupy_ref.Custom_Cupy_Ref(ptr = a.data.ptr, size = a.size)
-    c = gpuMemManagement.test_copy_custom_cupy_of_ones_to_cpu(b)
+    c = gpuMemManagement.test_copy_custom_cupy_to_cpu(b)
 
     assert(c == 1)
 
-#test 8 : copy array of float from custom cupy to cpu
-def test_copy_custom_cupy_of_float_to_cpu():
-    a = cp.array([3.14, 3.14, 3.14])
+#test 8 : test if real cupy can run successfully in cuda kernel
+def test_real_cupy_pointer_with_cuda_kernel():
+    a = cp.array([4.5, 4.5, 4.5])
+    b = gpuMemManagement.real_cupy_increment_all_data_by_1(a.data.ptr, a.size)
+
+    assert(b == 3*(4.5 + 1))
+
+#test 9 : test if custom cupy can run successfully in cuda kernel
+def test_custom_cupy_pointer_with_cuda_kernel():
+    a = cp.array([4.5, 4.5, 4.5])
     b = cupy_ref.Custom_Cupy_Ref(ptr = a.data.ptr, size = a.size)
-    c = gpuMemManagement.test_copy_custom_cupy_of_float_to_cpu(b)
+    c = gpuMemManagement.custom_cupy_increment_all_data_by_1(b)
 
-    assert(c == 1)
+    assert(c == 3*(4.5 + 1))
 
-#test 9 : test 2 reinterpret cast with different data type and see if both has a same value
-#this part need more explanation
-def test_2_different_reiterpret_cast():
-    a = cp.array([3.14, 3.14, 3.14])
-    b = gpuMemManagement.test_2_different_reiterpret_cast(a.data.ptr)
+#one of the reason why test 1 - 9 was successfull is because all cupy array value in the test is always double precision python float,
+#the reinterpret cast of the real cupy is always casted to double, and the custom cupy current design is only accepting double type.
+#test 10 : what if instead of python float, array of integer is sent by real cupy. expectation : failure
```

Comment:
That's make totally sense. By default numpy uses `numpy.float64`, if you create a new array with python floating pointer numbers, which is equal to C++ `double`. If you want to create a numpy array with `numpy.float32` (which is equal to C++ `float`), you have to set the element type manual: `a = np.array([3.0, 4.5], dtype=np.float32)` 

Example: https://github.com/ComputationalRadiationPhysics/student_project_python_bindings/blob/39c854178a1f7717667f17c41f6946347bb3496f/cupy_array_caster/test/0004_pointer_caster/test_pointerCaster.py#L77